Sources/CacheAdvance/CacheReader.swift

```diff
@@ -44,29 +41,26 @@ final class CacheReader {
         reader.offsetInFile
     }
 
-    /// Returns the next encodable message, seeking to the beginning of the next message.
-    func nextEncodedMessage() throws -> Data? {
-        let startingOffset = offsetInFile
-
-        guard startingOffset != offsetInFileAtEndOfNewestMessage else {
-            // We're at the last message.
-            return nil
+    /// Returns the encodable messages in a range
+    func encodedMessagesFromOffset(_ startOffset: UInt64, endOffset: UInt64? = nil) throws -> [Data] {
+        var encodedMessages = [Data]()
+        try reader.seek(to: startOffset)
+        while let data = try nextEncodedMessage() {
+            encodedMessages.append(data)
+            if let endOffset, offsetInFile >= endOffset {
+                break
+            }
         }
+        if let endOffset, offsetInFile != endOffset {
+            throw CacheAdvanceError.fileCorrupted
+        }
+        return encodedMessages
+    }
 
+    /// Returns the next encodable message, seeking to the beginning of the next message.
+    func nextEncodedMessage() throws -> Data? {
         switch try nextEncodedMessageSpan() {
         case let .span(messageLength):
-            // Check our assumptions before we try to read the message.
-            let endOfMessage = startingOffset + UInt64(MessageSpan.storageLength) + UInt64(messageLength)
-            let startingOffsetIsBeforeEndOfNewestMessageAndDoesNotExceedEndOfNewestMessage = startingOffset < offsetInFileAtEndOfNewestMessage && endOfMessage <= offsetInFileAtEndOfNewestMessage
-            let startingOffsetIsAfterEndOfNewestMessageAndDoesNotExceedEndOfFile = offsetInFileAtEndOfNewestMessage < startingOffset && endOfMessage <= maximumBytes
-            guard
-                startingOffsetIsBeforeEndOfNewestMessageAndDoesNotExceedEndOfNewestMessage
-                    || startingOffsetIsAfterEndOfNewestMessageAndDoesNotExceedEndOfFile
-            else {
-                // The offsetInFileAtEndOfNewestMessage is incorrect. This likely occured due to a crash when writing our header file.
-                throw CacheAdvanceError.fileCorrupted
-            }
```

Comment:
can you help me understand why we no longer need these checks?

Reply:
I would like to say we are safe if do not seek back in this method.

Reply:
Got it. Because these guards are preventing the seeking behavior which we are no longer doing we don't need them.  🤔. This makes sense. Will dig in more next week!

Reply:
Wish you have a nice weekend! and continue the discussion next week : )